From bfee5ae859fddd8e513aa22860b59d7300e00bad Mon Sep 17 00:00:00 2001
From: gbhakteX <gurudattax.bhakte@intel.com>
Date: Fri, 14 Sep 2012 11:30:43 +0530
Subject: [PORT FROM MAIN]Audio-MW: Some AVI clip (that include AAC, AAC-HEv1, AAC-HEv2 audio) can't           play good from gallery.

BZ: 47097

Present AVIExtractor doesn't support AAC audio.
With this fix, added support for AAC audio (WaveFormat = 0xFF) in AVI container.

Change-Id: Ibb716498c03d7aa235b9826d8023d3d3ac8843c5
Orig-Change-Id: Id9a02a9dd89fac7094158333a442325b9b199919
Signed-off-by: gbhakteX <gurudattax.bhakte@intel.com>
---
 media/libstagefright/AVIExtractor.cpp       |   19 ++++++++++++++++++-
 media/libstagefright/include/AVIExtractor.h |    4 ++++
 2 files changed, 22 insertions(+), 1 deletions(-)

diff --git a/media/libstagefright/AVIExtractor.cpp b/media/libstagefright/AVIExtractor.cpp
index e48e9f8..ff6c1b7 100644
--- a/media/libstagefright/AVIExtractor.cpp
+++ b/media/libstagefright/AVIExtractor.cpp
@@ -711,8 +711,25 @@ status_t AVIExtractor::parseStreamFormat(off64_t offset, size_t size) {
     } else {
         uint32_t format = U16LE_AT(data);
 
-        if ((format == 0x55) || (format == 0x50)) {
+        if ((format == WAVE_FORMAT_MPEGLAYER3) || (format == WAVE_FORMAT_MPEG)) {
             track->mMeta->setCString(kKeyMIMEType, MEDIA_MIMETYPE_AUDIO_MPEG);
+        } else if (format == WAVE_FORMAT_AAC) {
+//          The WAVEFORMATEX structure in AVI container is as below:
+//          ____________________________________________________________________________
+//          |Format|channel|sample rate|nAvgBytesPerSec|nBlockAlign|wBitsPerSample|cbSize|
+//     Bytes 0    1  2   3   4 ----- 7   8   -----  11   12 --- 13    14 --- 15    16  17
+//          where Format = 0x00FF, for AAC Audio.
+//          After these 18 bytes the bitstream contains the following information:
+//          ______Byte[18]_____,___Byte[19]______
+//          |AACProfile+1 |sf_index |ChannelCount| .....
+//     Bits  7   -----  3, 2--0 | 7,  6 ----- 3 , .....
+
+            uint8_t profile, sf_index, channels;
+            profile = (data[18] >> 3) -1;
+            sf_index = (((data[18] & 0x7) << 1 ) | (data[19] >> 0x7));
+            channels = (data[19] & 0x7F) >> 3;
+            track->mMeta = MakeAACCodecSpecificData(profile, sf_index, channels);
+            return OK;
         } else {
             ALOGW("Unsupported audio format = 0x%04x", format);
         }
diff --git a/media/libstagefright/include/AVIExtractor.h b/media/libstagefright/include/AVIExtractor.h
index ff5dcb5..bd5a1b1 100644
--- a/media/libstagefright/include/AVIExtractor.h
+++ b/media/libstagefright/include/AVIExtractor.h
@@ -25,6 +25,10 @@
 
 namespace android {
 
+#define WAVE_FORMAT_MPEG (0x0050)
+#define WAVE_FORMAT_MPEGLAYER3 (0x0055)
+#define WAVE_FORMAT_AAC (0x00FF)
+
 struct AVIExtractor : public MediaExtractor {
     AVIExtractor(const sp<DataSource> &dataSource);
 
-- 
1.7.4.1

